### ISSUE TITLE
Linux phone version is crashing

### ISSUE THREAD

**Issue by programmin1** [organicmaps/organicmaps#2623]
I was excited to see a Linux app for offline mapping, and on the flathub that installs on Librem phone! 
https://howtotrainyourrobot.com/snap-store-and-flatpak-missing-app-stores-for-librem-5-phone/
Unfortunately it doesn't load, I once got it to show the initial screen but now it doesn't load at all.
Also it only shows when I choose show non mobile-apps. Can it be optimized for everyone on Librem5, Pinephone and similar Linux phones that are gaining popularity these days?
And is there a build guide for the Linux version where one might try to see what is going wrong with this to fix it?

**Comment by biodranik** [comment]
Did you check our readme and the docs folder? There are detailed instructions on build.

I assume that your issue is similar to this one: https://github.com/flathub/app.organicmaps.desktop/issues/1

**Comment by Ferenc-** [comment]
I think the title is a bit confusing, because if one is running X11 on a phone instead of wayland, then it is not crashing, and considering it really is about the flathub available package, that is not a "linux phone version" but a linux desktop version running on a phone.

> Also it only shows when I choose show non mobile-apps. 

Well `app.organicmaps.desktop` is as one might expect is targeting the desktop not the mobile,
and that is exatly what I personally would expect, that it only shows up in phosh if one select so show the non mobile-apps.

> And is there a build guide for the Linux version where one might try to see what is going wrong with this to fix it?

If you are talking about the flatpak, then does that mean that you can't build from the flatpak manifest?
I am quite confident that it is buildable. Now regardless of that, I think if you have an issue with the flatpak manifest, then it's probably better to open an issue [where the manifest is](https://github.com/flathub/app.organicmaps.desktop).
If you meant a non-flatpak build, then depending on the distro, it might have been [discussed here](https://github.com/organicmaps/organicmaps/discussions/375#discussioncomment-2306536) already?

> Unfortunately it doesn't load, I once got it to show the initial screen but now it doesn't load at all.

That does sound like the previously linked https://github.com/flathub/app.organicmaps.desktop/issues/1,
where we already have more relevant info gathered. So I think this issue is a duplicate of that.

**Comment by biodranik** [comment]
@programmin1 can you please build and test version with these changes? #2947
@Ferenc- is it possible to build a binary for Linux from the branch and share it for tests?

**Comment by Ferenc-** [comment]
Hi, sure it is possible.
Only I am afraid it is a bit premature, that PR would need something like a "Needs help" label,
as there are still crashes on some ES2 hardware, on top of the blurryness etc.

**Comment by biodranik** [comment]
Please share crash logs, we can help.

**Comment by boroli** [comment]
That's the output of version 2022.08.23-5 running (well try to start) on mobian (phosh) on a PinePhone:

```
LOG TID(1) INFO   8.5251e-05 platform/string_storage_base.cpp:26 StringStorageBase(): Settin
gs path: /home/mobian/.var/app/app.organicmaps.desktop/config/OMaps/settings.ini
LOG TID(1) INFO     0.244494 storage/storage.cpp:638 LoadCountriesFile(): Loaded countries l
ist for version: 220816
LOG TID(1) INFO      1.14175 search/processor.cpp:215 SetPreferredLocale(): New preferred lo
cale: en
LOG TID(2) WARNING   1.17741 search/geocoder.cpp:726 CacheWorldLocalities(): Can't find Worl
d map file.
LOG TID(2) WARNING    1.1776 search/cities_boundaries_table.cpp:49 Load(): Can't find World 
map file.
LOG TID(2) WARNING   1.17763 search/processor.cpp:341 LoadCitiesBoundaries(): Can't load cit
ies boundaries
LOG TID(1) INFO      1.43774 storage/storage.cpp:326 RegisterAllLocalMaps(): Found file: Wor
ld in directory: 
LOG TID(1) INFO       1.4379 storage/storage.cpp:326 RegisterAllLocalMaps(): Found file: Wor
ldCoasts in directory: 
LOG TID(1) INFO      1.43981 map/framework.cpp:146 RegisterMap(): Loaded World map, of versi
on 220816
LOG TID(1) INFO      1.44043 map/framework.cpp:146 RegisterMap(): Loaded WorldCoasts map, of
 version 220816
LOG TID(1) INFO      1.46002 map/framework.cpp:377 Framework(): System languages: en_GB.UTF-
8
LOG TID(1) WARNING   1.46098 power_management/power_manager.cpp:71 Load(): Cannot read power
 manager config file. Exception: /home/mobian/.var/app/app.organicmaps.desktop/config/OMaps/
power_manager_config; Read; No such file or directory
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
Cannot find EGLConfig, returning null config
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
QOpenGLWidget: Failed to create context
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
QOpenGLWidget: Failed to create context
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
QOpenGLWidget: Failed to create context
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
qt.qpa.backingstore: composeAndFlush: QOpenGLContext creation failed
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
Cannot find EGLConfig, returning null config
```

I'm aware that the packaging was done in mind for desktop X11 setups. Nevertheless, it would be nice to get OrganicMaps also running on mobile devices under wayland.

**Comment by biodranik** [comment]
Can you build and test Linux mobile version locally? In theory, it should work if the same mobile OpenGL ES code that is used for Android will be reused for Linux. It should be easy to add/fix appropriate defines and build a separate mobile version for Linux.

Building a universal binary may be a more complex task.

**Comment by CaptainMorgan12** [comment]
i have the same issue on L5, flatpak install, wayland, pureos 10, GTK phosh shell 0.22.0, organicmaps v 2022.11.02, qtwayland 5.15.2.

>  flatpak run app.organicmaps.desktop
QIBusPlatformInputContext: invalid portal bus.
QSocketNotifier: Can only be used with threads started with QThread
LOG TID(1) INFO    5.136e-05 platform/string_storage_base.cpp:26 StringStorageBase(): Settings path: /home/purism/.var/app/app.organicmaps.desktop/config/OMaps/settings.ini                          
LOG TID(1) INFO     0.174545 storage/storage.cpp:624 LoadCountriesFile(): Loaded countries list for version: 221029                 
LOG TID(1) INFO      0.96454 search/processor.cpp:215 SetPreferredLocale(): New preferred locale: en
LOG TID(2) WARNING  0.992043 search/geocoder.cpp:736 CacheWorldLocalities(): Can't find World map file.
LOG TID(2) WARNING  0.992171 search/cities_boundaries_table.cpp:49 Load(): Can't find World map file.
LOG TID(2) WARNING  0.992199 search/processor.cpp:341 LoadCitiesBoundaries(): Can't load cities boundaries
LOG TID(1) INFO      1.17836 storage/storage.cpp:959 RegisterLocalFile(): Found file: World in directory:  with size: 40185789
LOG TID(1) INFO      1.17848 storage/storage.cpp:959 RegisterLocalFile(): Found file: WorldCoasts in directory:  with size: 8748723
LOG TID(1) INFO      1.17929 map/framework.cpp:147 RegisterMap(): Loaded World map, of version 221029
LOG TID(1) INFO      1.17952 map/framework.cpp:147 RegisterMap(): Loaded WorldCoasts map, of version 221029
LOG TID(1) INFO      1.19255 map/framework.cpp:378 Framework(): System languages: en_US.UTF-8
LOG TID(1) WARNING   1.19308 power_management/power_manager.cpp:71 Load(): Cannot read power manager config file. Exception: /home/purism/.var/app/app.organicmaps.desktop/config/OMaps/power_manager_config; Read; No such file or directory
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
libpng warning: iCCP: known incorrect sRGB profile
Cannot find EGLConfig, returning null config
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
QOpenGLWidget: Failed to create context
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
QOpenGLWidget: Failed to create context
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
QOpenGLWidget: Failed to create context
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
qt.qpa.backingstore: composeAndFlush: QOpenGLContext creation failed
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed
Cannot find EGLConfig, returning null config
QWaylandGLContext: failed to create EGLContext, error=3009
qt.qpa.backingstore: composeAndFlush: makeCurrent() failed

**Comment by boroli** [comment]
> Can you build and test Linux mobile version locally? In theory, it should work if the same mobile OpenGL ES code that is used for Android will be reused for Linux. It should be easy to add/fix appropriate defines and build a separate mobile version for Linux.

I tried building a locally Linux mobile version, following [these steps](https://wiki.mobian-project.org/doku.php?id=organicmaps#compiling-from-source). But as the my phone has only 2GB memory, and not 20GB free storage space, it's actually not possible to build organicmaps directly on the phone. Thus, I built it via QEMU in an emulated image, on a amd64 host, as I wasn't aware if cross-compilation would be enough.

In any case, after copying the build directory including the binary, it started only once. I accepted the copyright, and also accepted to download the world map. This eventually lead to crash Phosh. After Phosh restarted, I tried again starting OMaps, but then I ended up having the same error message, as with the flatpak binary.
 
> Building a universal binary may be a more complex task.

That's fair enough, and fully understandable.

**Comment by sinashaw** [comment]
> QWaylandGLContext: failed to create EGLContext, error=3009

In 2022, I had a 3009 problem, which has been solved. It is related to the system identification of OPENGLES version. When I have time, I may submit a revised version.

**Comment by Ferenc-** [comment]
@sinashaw I don't think it brings the cause forward to echo the current state of `master` yet again.
IMHO it would be time better spent for everyone, if you built and tried [this PR branch](https://github.com/organicmaps/organicmaps/pull/2947), with the currently needed `--es2` switch.

**Comment by sinashaw** [comment]
> Hi, sure it is possible.
> Only I am afraid it is a bit premature, that PR would need something like a "Needs help" label,
> as there are still crashes on some ES2 hardware, on top of the blurryness etc.

Mobile phone hardware supports es3, but our software thinks it is es2
